staging/src/kubevirt.io/client-go/apis/snapshot/v1alpha1/types.go

```diff
@@ -203,3 +203,79 @@ type VolumeSnapshotStatus struct {
 	// +optional
 	Error *VirtualMachineSnapshotError `json:"error,omitempty"`
 }
+
+// VirtualMachineRestore defines the operation of restoring a VM
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type VirtualMachineRestore struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec VirtualMachineRestoreSpec `json:"spec"`
+
+	// +optional
+	Status *VirtualMachineRestoreStatus `json:"status,omitempty"`
+}
+
+// VirtualMachineRestoreSpec is the spec for a VirtualMachineRestoreresource
+type VirtualMachineRestoreSpec struct {
+	// initially only VirtualMachine type supported
+	Target corev1.TypedLocalObjectReference `json:"target"`
+
+	VirtualMachineSnapshotName string `json:"virtualMachineSnapshotName"`
+
+	// only one of IncludeVolumes/ExcludeVolumes may be specified
+	// +optional
+	IncludeVolumes []string `json:"includeVolumes,omitempty"`
+
+	// +optional
+	ExcludeVolumes []string `json:"excludeVolumes,omitempty"`
```

Comment:
why do we need these options? What would be some example use cases?

Reply:
say one volume has your data and the other has the system/application.  You want to rollback a system update but not the data or vice/versa.

Reply:
I kind of get what you're saying, but the mechanics of how this work are a little fuzzy. I think i'm holding on to some assumptions that might not align exactly with what you;re going for with restore.

Have you documented any full snapshot and restore workflows we can review. I think that would help clarify (at least to me) what's required by the API.

Reply:
This all seemed pretty straightforward to me until I actually thought about it, lol.

I think it actually is very simple if we assume that volumes are never added/removed form the VM.

If neither include/excludeVolumes is specified, all eligible disks are restored.

If both include/excludeVolumes are specified, that's an error.

`includeVolumes` will only restore the specified volumes.

`excludeVolumes` will restore all eligible volumes but those specified.

BUT what if a snapshot contains a volume that was removed?  I guess we'll add it back?

OR what if a volume was added to a VM after the snapshot?  Do we keep it?  Remove it?  Actually delete it?  If so, I hope you took a snapshot before the restore.  Maybe we should always do that anyway.

Is this the sort of stuff you were wondering about @davidvossel?

Reply:
I also initially thought that include/excludeVolumes would be populated with volumes currently in the vm.  But may have to actually be volumes in the target snapshot

Reply:
Hmm.  The more I think about this, the more I think that (for at least the first pass of restore) all volumes that were part of the snapshot should be part of the restore.  If someone wants to do something tricky they could clone a new volume from one of the disk snapshots and attach that to the VM.  I think we should make the API as simple as possible and only add more functionality if it is truly needed.

Reply:
yeah let's skip for now

Reply:
> The more I think about this, the more I think that (for at least the first pass of restore) all volumes that were part of the snapshot should be part of the restore. 

yep, these summarize my thoughts as well. include/exclude volumes might make sense, but I'm not sure I want to get hung up on that until we get the basics in place.